### ISSUE TITLE
APIs like remove_nodes() and remove_edges() leave invalid cached batch info

### ISSUE THREAD

**Issue by maxwells-daemons** [dmlc/dgl#2310]
## 🐛 Bug

Invoking `remove_nodes(...)` or `remove_edges(...)` on a batched `DGLHeteroGraph`, and then attempting to unbatch the graph, causes an error. This is because `remove_nodes()` and `remove_edges()` do not correctly reset cached state.  

## To Reproduce

Here is a minimal reproducible example:
```
import dgl

g1 = dgl.graph(([0], [1]))
g2 = dgl.graph(([0], [1]))
batch = dgl.batch([g1, g2])
batch.remove_nodes([0])  # Or `batch.remove_edges([0])`
dgl.unbatch(batch)
```
This produces the following stack trace: 

```
Using backend: pytorch
Traceback (most recent call last):
  File "bug_repro_node.py", line 7, in <module>
    dgl.unbatch(batch)
  File "/home/whillikers/.cache/pypoetry/virtualenvs/compositional-graph-models-0kbwFTzO-py3.8/lib/python3.8/site-packages/dgl/batch.py", line 377, in unbatch
    us = F.split(u, edge_split[rel], 0)
  File "/home/whillikers/.cache/pypoetry/virtualenvs/compositional-graph-models-0kbwFTzO-py3.8/lib/python3.8/site-packages/dgl/backend/pytorch/tensor.py", line 168, in split
    return th.split(input, sizes_or_sections, dim)
  File "/home/whillikers/.cache/pypoetry/virtualenvs/compositional-graph-models-0kbwFTzO-py3.8/lib/python3.8/site-packages/torch/functional.py", line 115, in split
    return tensor.split(split_size_or_sections, dim)
  File "/home/whillikers/.cache/pypoetry/virtualenvs/compositional-graph-models-0kbwFTzO-py3.8/lib/python3.8/site-packages/torch/tensor.py", line 379, in split
    return super(Tensor, self).split_with_sizes(split_size, dim)
RuntimeError: start (1) + length (1) exceeds dimension size (1).
```

~However, inserting `batch._reset_cached_info()` before unbatching causes the correct result.~

EDIT: After calling `batch._reset_cached_info()`, the batch incorrectly reports a `batch_size` of 1. Currently I do not believe there's a way to remove nodes or edges from a batched graph and correctly retain or reconstruct the batch info. 

## Expected behavior

Calling `remove_nodes()` or `remove_edges()` on a graph should invalidate their cached batch information because their structure has changed.

## Environment

DGL version 0.5.2.

**Comment by mufeili** [comment]
I think this is a duplicate of [this discussion thread](https://discuss.dgl.ai/t/how-to-remove-nodes-from-batched-graph/1439). See if it addresses your issues.

**Comment by maxwells-daemons** [comment]
Yep, that looks like a duplicate. Thanks for the response -- I'll leave this open in case it's helpful, but feel free to close.

**Comment by jermainewang** [comment]
We should fix this. APIs that change graph topology should reset the batch information to default.

**Comment by NiklasBeierl** [comment]
I have a big graph from which I am removing nodes due to what I described in #3003.
When applying the workaround I described there, I found that I need to call `remove_edges` twice in order for it to take effect.

![Peek 2021-06-10 11-14](https://user-images.githubusercontent.com/22919961/121498785-0e4fd380-c9dd-11eb-8798-e06e476daf23.gif)


![Peek 2021-06-10 11-13](https://user-images.githubusercontent.com/22919961/121498606-dcd70800-c9dc-11eb-8e4c-3b33d517d6b1.gif)

Is this the same issue?

**Comment by mufeili** [comment]
[DGLGraph.edges](https://docs.dgl.ai/generated/dgl.DGLGraph.edges.html#dgl.DGLGraph.edges) by default returns the source and destination nodes of the edges while [DGLGraph.remove_edges](https://docs.dgl.ai/generated/dgl.DGLGraph.remove_edges.html#dgl.DGLGraph.remove_edges) takes edge ids as input. So what you are looking for is in fact batch_graph.remove_edges(torch.arange(batch_graph.num_edges('is'))).

**Comment by jermainewang** [comment]
Fixed in #3119. The patch will be in the next 0.7 release.
